rmp-serde/src/encode.rs

```diff
@@ -167,109 +159,112 @@ impl<'a, W: VariantWriter> Serializer<'a, W> {
             depth: 1000,
         }
     }
-
-    #[inline]
-    fn serialize_variant<F>(&mut self, variant_index: usize, maybe_len: Option<usize>, mut visit: F) -> Result<(), Error>
-        where F: FnMut(&mut Self) -> Result<Option<()>, <Self as serde::Serializer>::Error>
-    {
-        try!(write_array_len(&mut self.wr, 2));
-
-        // Encode a value position...
-        try!(serde::Serializer::serialize_usize(self, variant_index));
-
-        let len = match maybe_len {
-            Some(len) => len,
-            None => return Err(Error::UnknownLength),
-        };
-
-        // ... and its arguments length.
-        try!(write_array_len(&mut self.wr, len as u32));
-
-        while let Some(()) = try!(depth_count!(self.depth, visit(self))) { }
-
-        Ok(())
-    }
 }
 
 impl<'a, W: VariantWriter> serde::Serializer for Serializer<'a, W> {
     type Error = Error;
 
+    type SeqState = ();
+    type TupleState = ();
+    type TupleStructState = ();
+    type TupleVariantState = ();
+    type MapState = ();
+    type StructState = ();
+    type StructVariantState = ();
+
+    #[inline]
     fn serialize_unit(&mut self) -> Result<(), Error> {
         write_nil(&mut self.wr).map_err(From::from)
     }
 
+    #[inline]
```

Comment:
Is such hint proved with benchmarks/assembly? I think the compiler is able to optimize one-liners during LTO easily even without this hint.